sdk/core/azure-core-amqp/src/amqp/message_receiver.cpp

```diff
@@ -541,9 +578,15 @@ namespace Azure { namespace Core { namespace Amqp { namespace _detail {
       bool shouldWaitForClose = m_currentState == _internal::MessageReceiverState::Closing
           || m_currentState == _internal::MessageReceiverState::Open;
 
-      Common::_detail::GlobalStateHolder::GlobalStateInstance()->RemovePollable(
-          m_link); // This will ensure that the link is cleaned up on the next poll()
-
+      {
+        std::unique_lock<std::mutex> lock{m_MutableState};
+        if (m_linkPollingEnabled)
+        {
+          Common::_detail::GlobalStateHolder::GlobalStateInstance()->RemovePollable(
```

Comment:
It's under a lock, and this is the only code which calls RemovePollable. How would moving the linkPollingEnabled = false before the RemovePollable fix anything?

Reply:
Should this get called from a separate thread (which it probably shouldn't based on your question), an if that happens while in `RemovePollable()` then execution of another thread will once again enter inside `if (m_linkPollingEnabled)`, however if we set it to false before calling, there is less chance of this happening. Okay, I realized, even then it is too fragile without atomics and mutexes, and we'd need more than what I am suggesting to enable it, and this must never be getting called concurrently. So I retract my suggestion.

Reply:
??? There *is* a mutex here - the `m_mutableState` mutex, and it's held as of line 582. 

I originally wrote the code without the mutex and then realized that I had a concurrency problem, so I added the m_mutableState mutex to cover this case. 

Currently the only state covered by the `m_mutableState` mutex is the `m_linkPollingEnabled` field in the struct. 

Note that there *is* a possible deadlock here, because AddPollable and RemovePollable both take the global pollables lock, AND the RemovePollable blocks waiting on any polling operations to complete. So if you attempt to call AddPollable or RemovePollable while holding any common locks, you can have a deadlock (in fact, that's *exactly* the reason that the `EnableLinkPolling` method was added - we had a case where we could call `AddPollable` while holding the connection lock which triggeed a deadlock.